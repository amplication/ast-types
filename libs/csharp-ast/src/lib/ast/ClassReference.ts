import { AstNode } from "../core/AstNode";
import { Writer } from "../core/Writer";

/**
 * Configuration arguments for creating a C# class reference.
 */
export declare namespace ClassReference {
  interface Args {
    /** The name of the C# class */
    name: string;
    /** The namespace of the C# class */
    namespace: string;
  }
}

/**
 * Represents a reference to a C# class in the AST.
 * This class handles the generation of C# class references including
 * the class name and its namespace.
 *
 * @extends {AstNode}
 */
export class ClassReference extends AstNode {
  /** The name of the class */
  public readonly name: string;
  /** The namespace containing the class */
  public readonly namespace: string;

  /**
   * Creates a new C# class reference.
   * @param {ClassReference.Args} args - The configuration arguments for the class reference
   */
  constructor({ name, namespace }: ClassReference.Args) {
    super();
    this.name = name;
    this.namespace = namespace;
  }

  /**
   * Writes the class reference to the writer.
   * @param {Writer} writer - The writer to write to
   */
  public write(writer: Writer): void {
    writer.addReference(this);
    writer.write(`${this.name}`);
  }
}

/**
 * Pre-defined class reference for the OneOf type.
 */
export const OneOfClassReference = new ClassReference({
  name: "OneOf",
  namespace: "OneOf",
});

//   TODO: remove this in favor of the one in PrebuiltUtilities

/**
 * Pre-defined class reference for the StringEnum type.
 */
export const StringEnumClassReference = new ClassReference({
  name: "StringEnum",
  namespace: "StringEnum",
});
