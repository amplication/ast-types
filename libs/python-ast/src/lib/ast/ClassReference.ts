import { AstNode } from "../core/AstNode";
import { Writer } from "../core/Writer";

/**
 * Configuration arguments for creating a Python class reference.
 */
export interface ClassReferenceArgs {
  /** The name of the class */
  name: string;
  /** The module name containing the class */
  moduleName?: string;
  /** Optional alias for the class when imported */
  alias?: string;
}

/**
 * Represents a reference to a Python class.
 * This class is used to track and generate imports for classes used in the code.
 *
 * @extends {AstNode}
 */
export class ClassReference extends AstNode {
  /** The name of the class */
  public readonly name: string;
  /** The module name containing the class */
  public readonly moduleName?: string;
  /** Optional alias for the class */
  public readonly alias?: string;

  /**
   * Creates a new Python class reference.
   * @param {ClassReferenceArgs} args - The configuration arguments
   */
  constructor({ name, moduleName, alias }: ClassReferenceArgs) {
    super();
    this.name = name;
    this.moduleName = moduleName;
    this.alias = alias;
  }

  /**
   * Writes the class reference to the writer and adds it to the import collection.
   * @param {Writer} writer - The writer to write to
   */
  public write(writer: Writer): void {
    if (this.moduleName) {
      writer.addImport(this);
    }
    writer.write(this.alias || this.name);
  }
}
