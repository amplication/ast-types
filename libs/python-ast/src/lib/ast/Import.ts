import { AstNode } from "../core/AstNode";
import { Writer } from "../core/Writer";

/**
 * Configuration arguments for creating a Python import statement.
 */
export interface ImportArgs {
  /** The module name to import */
  moduleName: string;
  /** The names to import from the module (if using 'from ... import ...') */
  names?: string[];
  /** The alias for the module (if using 'import ... as ...') */
  alias?: string;
}

/**
 * Represents a Python import statement.
 * This class handles the generation of Python import statements, including
 * simple imports, from imports, and imports with aliases.
 *
 * @extends {AstNode}
 */
export class Import extends AstNode {
  /** The module name to import */
  public readonly moduleName: string;
  /** The names to import from the module */
  public readonly names: string[];
  /** The alias for the module */
  public readonly alias?: string;

  /**
   * Creates a new Python import statement.
   * @param {ImportArgs} args - The configuration arguments
   */
  constructor({ moduleName, names = [], alias }: ImportArgs) {
    super();
    this.moduleName = moduleName;
    this.names = names;
    this.alias = alias;
  }

  /**
   * Writes the import statement to the writer.
   * @param {Writer} writer - The writer to write to
   */
  public write(writer: Writer): void {
    writer.addImport(this);
  }
}
