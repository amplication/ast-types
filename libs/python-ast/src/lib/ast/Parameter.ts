import { AstNode } from "../core/AstNode";
import { Writer } from "../core/Writer";
import { ClassReference } from "./ClassReference";

/**
 * Configuration arguments for creating a Python parameter.
 */
export interface ParameterArgs {
  /** The name of the parameter */
  name: string;
  /** The type annotation for the parameter */
  type?: ClassReference;
  /** The default value for the parameter */
  default_?: string;
  /** Whether this is a keyword-only parameter (after *) */
  isKeywordOnly?: boolean;
  /** Whether this is a positional-only parameter (before /) */
  isPositionalOnly?: boolean;
  /** Whether this is a variable positional parameter (*args) */
  isVariablePositional?: boolean;
  /** Whether this is a variable keyword parameter (**kwargs) */
  isVariableKeyword?: boolean;
}

/**
 * Represents a Python parameter in a function or method definition.
 * This class handles the generation of parameter declarations including
 * type annotations and default values.
 *
 * @extends {AstNode}
 */
export class Parameter extends AstNode {
  /** The name of the parameter */
  public readonly name: string;
  /** The type annotation for the parameter */
  public readonly type?: ClassReference;
  /** The default value for the parameter */
  public readonly default_?: string;
  /** Whether this is a keyword-only parameter */
  public readonly isKeywordOnly: boolean;
  /** Whether this is a positional-only parameter */
  public readonly isPositionalOnly: boolean;
  /** Whether this is a variable positional parameter */
  public readonly isVariablePositional: boolean;
  /** Whether this is a variable keyword parameter */
  public readonly isVariableKeyword: boolean;

  /**
   * Creates a new Python parameter.
   * @param {ParameterArgs} args - The configuration arguments
   */
  constructor({
    name,
    type,
    default_,
    isKeywordOnly = false,
    isPositionalOnly = false,
    isVariablePositional = false,
    isVariableKeyword = false,
  }: ParameterArgs) {
    super();
    this.name = name;
    this.type = type;
    this.default_ = default_;
    this.isKeywordOnly = isKeywordOnly;
    this.isPositionalOnly = isPositionalOnly;
    this.isVariablePositional = isVariablePositional;
    this.isVariableKeyword = isVariableKeyword;

    // Validate parameter configuration
    if (
      [
        isKeywordOnly,
        isPositionalOnly,
        isVariablePositional,
        isVariableKeyword,
      ].filter(Boolean).length > 1
    ) {
      throw new Error(
        "A parameter can only be one of: keyword-only, positional-only, variable positional, or variable keyword",
      );
    }
  }

  /**
   * Writes the parameter declaration to the writer.
   * @param {Writer} writer - The writer to write to
   */
  public write(writer: Writer): void {
    if (this.isVariablePositional) {
      writer.write("*");
    } else if (this.isVariableKeyword) {
      writer.write("**");
    }

    writer.write(this.name);

    if (this.type) {
      writer.write(": ");
      this.type.write(writer);
    }

    if (this.default_) {
      writer.write(" = ");
      writer.write(this.default_);
    }
  }
}
