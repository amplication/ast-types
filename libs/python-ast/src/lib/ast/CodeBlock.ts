import { AstNode } from "../core/AstNode";
import { Writer } from "../core/Writer";
import { Import } from "./Import";

/**
 * Configuration arguments for creating a Python code block.
 */
export type CodeBlockArgs =
  | {
      /** The code to write */
      code: string;
      /** A list of imports that are used in the code */
      imports?: Import[] | null;
    }
  | {
      /** A function that writes code to the writer */
      code: (writer: Writer) => void;
      imports?: never;
    };

/**
 * Represents a block of Python code in the AST.
 * This class handles the generation of arbitrary Python code blocks, either as
 * static strings or as dynamic code generation functions. It is used for language
 * features not directly supported by specific AST node classes.
 *
 * @extends {AstNode}
 */
export class CodeBlock extends AstNode {
  /** The code content or code generation function */
  private value: string | ((writer: Writer) => void);
  /** The imports used in the code */
  private imports: Import[];

  /**
   * Creates a new Python code block.
   * @param {CodeBlockArgs} args - The configuration arguments
   */
  constructor(args: CodeBlockArgs) {
    super();
    this.value = args.code;
    this.imports = [];

    if ("imports" in args && args.imports) {
      this.imports.push(...args.imports);
    }
  }

  /**
   * Writes the code block to the writer.
   * @param {Writer} writer - The writer to write to
   */
  public write(writer: Writer): void {
    if (typeof this.value === "string") {
      this.imports.forEach((import_) => writer.addImport(import_));
      writer.write(this.value);
    } else {
      this.value(writer);
    }
  }
}
