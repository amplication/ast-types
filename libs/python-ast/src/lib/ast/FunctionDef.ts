import { AstNode } from "../core/AstNode";
import { Writer } from "../core/Writer";
import { ClassReference } from "./ClassReference";
import { CodeBlock } from "./CodeBlock";
import { Decorator } from "./Decorator";
import { Parameter } from "./Parameter";
import { Return } from "./Return";

/**
 * Configuration arguments for creating a Python function.
 */
export interface FunctionDefArgs {
  /** The name of the function */
  name: string;
  /** The parameters of the function */
  parameters?: Parameter[];
  /** Decorators to apply to the function */
  decorators?: Decorator[];
  /** Documentation string for the function */
  docstring?: string;
  /** Return type annotation */
  returnType?: ClassReference;
  /** Whether this is a static method */
  isStatic?: boolean;
  /** Whether this is a class method */
  isClassMethod?: boolean;
  /** Whether this is an async function */
  isAsync?: boolean;
}

/**
 * Represents a Python function definition in the AST.
 * This class handles the generation of Python function declarations including
 * parameters, decorators, and return type annotations.
 *
 * @extends {AstNode}
 */
export class FunctionDef extends AstNode {
  /** The name of the function */
  public readonly name: string;
  /** The parameters of the function */
  public readonly parameters: Parameter[];
  /** Decorators applied to the function */
  public readonly decorators: Decorator[];
  /** The documentation string */
  public readonly docstring?: string;
  /** The return type annotation */
  public readonly returnType?: ClassReference;
  /** Whether this is a static method */
  public readonly isStatic: boolean;
  /** Whether this is a class method */
  public readonly isClassMethod: boolean;
  /** Whether this is an async function */
  public readonly isAsync: boolean;

  /** The body of the function */
  private body: (CodeBlock | Return)[] = [];

  /**
   * Creates a new Python function definition.
   * @param {FunctionDefArgs} args - The configuration arguments
   */
  constructor({
    name,
    parameters = [],
    decorators = [],
    docstring,
    returnType,
    isStatic = false,
    isClassMethod = false,
    isAsync = false,
  }: FunctionDefArgs) {
    super();
    this.name = name;
    this.parameters = parameters;
    this.decorators = [...decorators];
    this.docstring = docstring;
    this.returnType = returnType;
    this.isStatic = isStatic;
    this.isClassMethod = isClassMethod;
    this.isAsync = isAsync;

    // Add appropriate decorators for static and class methods
    if (
      this.isStatic &&
      !this.decorators.some((d) => d.name === "staticmethod")
    ) {
      this.decorators.push(new Decorator({ name: "staticmethod" }));
    }
    if (
      this.isClassMethod &&
      !this.decorators.some((d) => d.name === "classmethod")
    ) {
      this.decorators.push(new Decorator({ name: "classmethod" }));
    }
  }

  /**
   * Adds a statement to the function body.
   * @param {CodeBlock | Return} statement - The statement to add
   */
  public addStatement(statement: CodeBlock | Return): void {
    this.body.push(statement);
  }

  /**
   * Writes the function definition and its body to the writer.
   * @param {Writer} writer - The writer to write to
   */
  public write(writer: Writer): void {
    // Write decorators
    this.decorators.forEach((decorator) => {
      decorator.write(writer);
      writer.newLine();
    });

    // Write function definition
    if (this.isAsync) {
      writer.write("async ");
    }
    writer.write(`def ${this.name}(`);

    // Write parameters
    this.parameters.forEach((param, index) => {
      param.write(writer);
      if (index < this.parameters.length - 1) {
        writer.write(", ");
      }
    });

    writer.write(")");

    // Write return type annotation if provided
    if (this.returnType) {
      writer.write(" -> ");
      this.returnType.write(writer);
    }

    writer.write(":");
    writer.newLine();
    writer.indent();

    // Write docstring if provided
    if (this.docstring) {
      writer.writeLine(`"""${this.docstring}"""`);
      if (this.body.length > 0) {
        writer.newLine();
      }
    }

    // Write function body
    if (this.body.length > 0) {
      this.body.forEach((statement) => {
        statement.write(writer);
        writer.newLine();
      });
    } else {
      // Empty function body needs a pass statement
      writer.writeLine("pass");
    }

    writer.dedent();
  }
}
