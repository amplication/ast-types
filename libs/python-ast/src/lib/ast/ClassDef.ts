import { AstNode } from "../core/AstNode";
import { Writer } from "../core/Writer";
import { Assign } from "./Assign";
import { ClassReference } from "./ClassReference";
import { CodeBlock } from "./CodeBlock";
import { Decorator } from "./Decorator";
import { FunctionDef } from "./FunctionDef";

/**
 * Configuration arguments for creating a Python class.
 */
export interface ClassDefArgs {
  /** The name of the class */
  name: string;
  /** The module name containing the class */
  moduleName?: string;
  /** The parent classes to inherit from */
  bases?: ClassReference[];
  /** Decorators to apply to the class */
  decorators?: Decorator[];
  /** Documentation string for the class */
  docstring?: string;
}

/**
 * Represents a Python class definition in the AST.
 * This class handles the generation of Python class declarations including methods,
 * attributes, and nested classes.
 *
 * @extends {AstNode}
 */
export class ClassDef extends AstNode {
  /** The name of the class */
  public readonly name: string;
  /** The module name containing the class */
  public readonly moduleName?: string;
  /** The parent classes to inherit from */
  public readonly bases: ClassReference[];
  /** The decorators applied to the class */
  public readonly decorators: Decorator[];
  /** The documentation string for the class */
  public readonly docstring?: string;
  /** The class reference for this class */
  public readonly reference: ClassReference;

  /** The methods defined in this class */
  private methods: FunctionDef[] = [];
  /** The attributes defined in this class */
  private attributes: Assign[] = [];
  /** The nested classes defined in this class */
  private nestedClasses: ClassDef[] = [];
  /** Custom code blocks for unsupported features */
  private codeBlocks: CodeBlock[] = [];

  /**
   * Creates a new Python class definition.
   * @param {ClassDefArgs} args - The configuration arguments
   */
  constructor({
    name,
    moduleName,
    bases = [],
    decorators = [],
    docstring,
  }: ClassDefArgs) {
    super();
    this.name = name;
    this.moduleName = moduleName;
    this.bases = bases;
    this.decorators = decorators;
    this.docstring = docstring;

    this.reference = new ClassReference({
      name: this.name,
      moduleName: this.moduleName,
    });
  }

  /**
   * Adds a method to the class.
   * @param {FunctionDef} method - The method to add
   */
  public addMethod(method: FunctionDef): void {
    this.methods.push(method);
  }

  /**
   * Adds an attribute to the class.
   * @param {Assign} attribute - The attribute to add
   */
  public addAttribute(attribute: Assign): void {
    this.attributes.push(attribute);
  }

  /**
   * Adds a nested class to this class.
   * @param {ClassDef} nestedClass - The nested class to add
   */
  public addNestedClass(nestedClass: ClassDef): void {
    this.nestedClasses.push(nestedClass);
  }

  /**
   * Adds a code block for unsupported features.
   * @param {CodeBlock} codeBlock - The code block to add
   */
  public addCodeBlock(codeBlock: CodeBlock): void {
    this.codeBlocks.push(codeBlock);
  }

  /**
   * Writes the class definition and its members to the writer.
   * @param {Writer} writer - The writer to write to
   */
  public write(writer: Writer): void {
    // Write decorators
    this.decorators.forEach((decorator) => {
      decorator.write(writer);
      writer.newLine();
    });

    // Write class definition
    writer.write(`class ${this.name}`);

    // Write base classes if any
    if (this.bases.length > 0) {
      writer.write("(");
      this.bases.forEach((base, index) => {
        base.write(writer);
        if (index < this.bases.length - 1) {
          writer.write(", ");
        }
      });
      writer.write(")");
    }

    writer.write(":");
    writer.newLine();
    writer.indent();

    // Write docstring if provided
    if (this.docstring) {
      writer.writeLine(`"""${this.docstring}"""`);
      if (
        this.attributes.length > 0 ||
        this.methods.length > 0 ||
        this.nestedClasses.length > 0 ||
        this.codeBlocks.length > 0
      ) {
        writer.newLine();
      }
    }

    // Write attributes
    if (this.attributes.length > 0) {
      this.attributes.forEach((attribute) => {
        attribute.write(writer);
        writer.newLine();
      });
      if (
        this.methods.length > 0 ||
        this.nestedClasses.length > 0 ||
        this.codeBlocks.length > 0
      ) {
        writer.newLine();
      }
    }

    // Write methods
    this.methods.forEach((method, index, array) => {
      method.write(writer);
      if (index < array.length - 1) {
        writer.newLine();
      }
    });

    if (
      this.methods.length > 0 &&
      (this.nestedClasses.length > 0 || this.codeBlocks.length > 0)
    ) {
      writer.newLine();
      writer.newLine();
    }

    // Write nested classes
    this.nestedClasses.forEach((nestedClass, index, array) => {
      nestedClass.write(writer);
      if (index < array.length - 1) {
        writer.newLine();
        writer.newLine();
      }
    });

    if (this.nestedClasses.length > 0 && this.codeBlocks.length > 0) {
      writer.newLine();
      writer.newLine();
    }

    // Write additional code blocks
    this.codeBlocks.forEach((codeBlock, index, array) => {
      codeBlock.write(writer);
      if (index < array.length - 1) {
        writer.newLine();
      }
    });

    // If the class is empty, add a pass statement
    if (
      this.attributes.length === 0 &&
      this.methods.length === 0 &&
      this.nestedClasses.length === 0 &&
      this.codeBlocks.length === 0 &&
      !this.docstring
    ) {
      writer.writeLine("pass");
    }

    writer.dedent();
  }
}
