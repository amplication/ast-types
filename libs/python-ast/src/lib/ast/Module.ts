import { AstNode } from "../core/AstNode";
import { Writer } from "../core/Writer";
import { ClassDef } from "./ClassDef";
import { CodeBlock } from "./CodeBlock";
import { FunctionDef } from "./FunctionDef";
import { Import } from "./Import";

/**
 * Configuration arguments for creating a Python module.
 */
export interface ModuleArgs {
  /** The name of the module */
  name: string;
  /** The docstring for the module */
  docstring?: string;
}

/**
 * Represents a Python module in the AST.
 * This class is the top-level container for Python code, including imports,
 * functions, classes, and global code.
 *
 * @extends {AstNode}
 */
export class Module extends AstNode {
  /** The name of the module */
  public readonly name: string;
  /** The docstring for the module */
  public readonly docstring?: string;

  /** The imports defined in this module */
  private imports: Import[] = [];
  /** The children nodes (functions, classes, code blocks) in order of addition */
  private children: (FunctionDef | ClassDef | CodeBlock)[] = [];

  /**
   * Creates a new Python module.
   * @param {ModuleArgs} args - The configuration arguments
   */
  constructor({ name, docstring }: ModuleArgs) {
    super();
    this.name = name;
    this.docstring = docstring;
  }

  /**
   * Adds an import statement to the module.
   * @param {Import} importStatement - The import statement to add
   */
  public addImport(importStatement: Import): void {
    this.imports.push(importStatement);
  }

  /**
   * Adds a function to the module.
   * @param {FunctionDef} functionDef - The function to add
   */
  public addFunction(functionDef: FunctionDef): void {
    this.children.push(functionDef);
  }

  /**
   * Adds a class to the module.
   * @param {ClassDef} classDef - The class to add
   */
  public addClass(classDef: ClassDef): void {
    this.children.push(classDef);
  }

  /**
   * Adds a global code block to the module.
   * @param {CodeBlock} codeBlock - The code block to add
   */
  public addCodeBlock(codeBlock: CodeBlock): void {
    this.children.push(codeBlock);
  }

  /**
   * Writes the module and its contents to the writer.
   * @param {Writer} writer - The writer to write to
   */
  public write(writer: Writer): void {
    // Write imports
    if (this.imports.length > 0) {
      this.imports.forEach((import_) => import_.write(writer));
      writer.newLine();
    }

    // Write module docstring if provided
    if (this.docstring) {
      writer.writeLine(`"""${this.docstring}"""`);
      writer.newLine();
    }

    // Write all children in order
    if (this.children.length > 0) {
      this.children.forEach((child, index) => {
        child.write(writer);
        if (index < this.children.length - 1) {
          writer.newLine();
          writer.newLine();
        }
      });
    }
  }
}
