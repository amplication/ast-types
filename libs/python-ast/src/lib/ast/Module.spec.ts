import { Module } from "./Module";
import { ClassDef } from "./ClassDef";
import { FunctionDef } from "./FunctionDef";
import { Parameter } from "./Parameter";
import { ClassReference } from "./ClassReference";
import { CodeBlock } from "./CodeBlock";
import { Import } from "./Import";

describe("Module", () => {
  it("should generate an empty module with docstring", () => {
    const module = new Module({
      name: "config",
      docstring: "Configuration module for the application",
    });

    expect(module.toString()).toMatchSnapshot();
    expect(module.toString()).toContain(
      '"""Configuration module for the application"""',
    );
  });

  it("should generate a module with imports", () => {
    const module = new Module({
      name: "utils",
    });

    module.addImport(
      new Import({
        moduleName: "os",
      }),
    );

    module.addImport(
      new Import({
        moduleName: "datetime",
        names: ["datetime", "timezone"],
      }),
    );

    expect(module.toString()).toMatchSnapshot();
    expect(module.toString()).toContain("import os");
    expect(module.toString()).toContain(
      "from datetime import datetime, timezone",
    );
  });

  it("should generate a module with functions", () => {
    const module = new Module({
      name: "math_utils",
    });

    const squareFunction = new FunctionDef({
      name: "square",
      parameters: [
        new Parameter({
          name: "x",
          type: new ClassReference({ name: "float" }),
        }),
      ],
      returnType: new ClassReference({ name: "float" }),
    });
    squareFunction.addStatement(new CodeBlock({ code: "return x * x" }));
    module.addFunction(squareFunction);

    const cubeFunction = new FunctionDef({
      name: "cube",
      parameters: [
        new Parameter({
          name: "x",
          type: new ClassReference({ name: "float" }),
        }),
      ],
      returnType: new ClassReference({ name: "float" }),
    });
    cubeFunction.addStatement(new CodeBlock({ code: "return x * x * x" }));
    module.addFunction(cubeFunction);

    expect(module.toString()).toMatchSnapshot();
    expect(module.toString()).toContain("def square(x: float) -> float:");
    expect(module.toString()).toContain("return x * x");
    expect(module.toString()).toContain("def cube(x: float) -> float:");
    expect(module.toString()).toContain("return x * x * x");
  });

  it("should generate a module with classes", () => {
    const module = new Module({
      name: "models",
    });

    const userClass = new ClassDef({
      name: "User",
      docstring: "User model",
    });
    module.addClass(userClass);

    const productClass = new ClassDef({
      name: "Product",
      docstring: "Product model",
    });
    module.addClass(productClass);

    expect(module.toString()).toMatchSnapshot();
    expect(module.toString()).toContain("class User:");
    expect(module.toString()).toContain('"""User model"""');
    expect(module.toString()).toContain("class Product:");
    expect(module.toString()).toContain('"""Product model"""');
  });

  it("should generate a module with global code", () => {
    const module = new Module({
      name: "settings",
    });

    module.addCodeBlock(
      new CodeBlock({
        code: 'DEBUG = True\nVERSION = "1.0.0"\nBASE_DIR = os.path.dirname(os.path.abspath(__file__))',
        imports: [new Import({ moduleName: "os", names: ["os"] })],
      }),
    );

    expect(module.toString()).toMatchSnapshot();
    expect(module.toString()).toContain("import os");
    expect(module.toString()).toContain("DEBUG = True");
    expect(module.toString()).toContain('VERSION = "1.0.0"');
    expect(module.toString()).toContain(
      "BASE_DIR = os.path.dirname(os.path.abspath(__file__))",
    );
  });

  it("should generate a complete module with all features and maintain order", () => {
    const module = new Module({
      name: "app",
      docstring: "Main application module",
    });

    // Add imports
    module.addImport(
      new Import({
        moduleName: "os",
      }),
    );
    module.addImport(
      new Import({
        moduleName: "logging",
      }),
    );

    // Add global code
    module.addCodeBlock(
      new CodeBlock({
        code: 'logger = logging.getLogger(__name__)\nAPP_NAME = "MyApp"\nVERSION = "1.0.0"',
      }),
    );

    // Add a utility function
    const configFunction = new FunctionDef({
      name: "configure_app",
      parameters: [
        new Parameter({
          name: "debug",
          type: new ClassReference({ name: "bool" }),
          default_: "False",
        }),
      ],
    });
    configFunction.addStatement(
      new CodeBlock({
        code: 'logger.info("Configuring app with debug=%s", debug)\nreturn {"debug": debug, "version": VERSION}',
      }),
    );
    module.addFunction(configFunction);

    // Add an application class
    const appClass = new ClassDef({
      name: "Application",
      docstring: "Main application class",
    });

    const initMethod = new FunctionDef({
      name: "__init__",
      parameters: [
        new Parameter({ name: "self" }),
        new Parameter({
          name: "config",
          type: new ClassReference({ name: "dict" }),
        }),
      ],
    });
    initMethod.addStatement(
      new CodeBlock({
        code: 'self.config = config\nlogger.info("Application initialized")',
      }),
    );
    appClass.addMethod(initMethod);

    const runMethod = new FunctionDef({
      name: "run",
      parameters: [new Parameter({ name: "self" })],
    });
    runMethod.addStatement(
      new CodeBlock({
        code: 'logger.info("Running application %s", APP_NAME)\nprint("Application running...")',
      }),
    );
    appClass.addMethod(runMethod);

    module.addClass(appClass);

    expect(module.toString()).toMatchSnapshot();

    // Additional test to verify order
    const output = module.toString();
    const codeBlockIndex = output.indexOf(
      "logger = logging.getLogger(__name__)",
    );
    const functionIndex = output.indexOf("def configure_app");
    const classIndex = output.indexOf("class Application:");

    expect(codeBlockIndex).toBeLessThan(functionIndex);
    expect(functionIndex).toBeLessThan(classIndex);
  });
});
