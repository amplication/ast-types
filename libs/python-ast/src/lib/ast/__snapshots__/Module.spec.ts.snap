// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Module should generate a complete module with all features 1`] = `
"import os
import logging


"""Main application module"""

logger = logging.getLogger(__name__)
APP_NAME = "MyApp"
VERSION = "1.0.0"

def configure_app(debug: bool = False):
    logger.info("Configuring app with debug=%s", debug)
    return {"debug": debug, "version": VERSION}
    

class Application:
    """Main application class"""
        
        def __init__(self, config: dict):
            self.config = config
        logger.info("Application initialized")
            
        def run(self):
            logger.info("Running application %s", APP_NAME)
        print("Application running...")
        "
`;

exports[`Module should generate a module with classes 1`] = `
"class User:
    """User model"""
    

class Product:
    """Product model"""
    "
`;

exports[`Module should generate a module with functions 1`] = `
"def square(x: float) -> float:
    return x * x
    

def cube(x: float) -> float:
    return x * x * x
    "
`;

exports[`Module should generate a module with global code 1`] = `
"from os import os

DEBUG = True
VERSION = "1.0.0"
BASE_DIR = os.path.dirname(os.path.abspath(__file__))
"
`;

exports[`Module should generate a module with imports 1`] = `
"import os
from datetime import datetime, timezone


"
`;

exports[`Module should generate an empty module with docstring 1`] = `
""""Configuration module for the application"""

"
`;
